File: py/day5.py
data = [[int(x) for x in row.replace(' -> ', ',').split(',')] for row in open('./data/5.txt').readlines()]
silver = {}
gold = {}

def line(graph, x1, y1, x2, y2):
    dx = 0 if x1 == x2 else (1 if x2 > x1 else -1)
    dy = 0 if y1 == y2 else (1 if y2 > y1 else -1)
    while x1 != x2 + dx or y1 != y2 + dy:
        key = str(x1) + ',' + str(y1)
        graph[key] = 1 if not key in graph else graph[key] + 1
        x1 += dx
        y1 += dy

for row in data:
    x1, y1, x2, y2 = row
    line(gold, x1, y1, x2, y2)
    if x1 == x2 or y1 == y2:
        line(silver, x1, y1, x2, y2)

print(sum([1 for key in silver if silver[key] > 1]))
print(sum([1 for key in gold if gold[key] > 1]))

File: js/day5.js
fetch('./data/5.txt')
  .then(res => res.text())
  .then(text => {
    let ctx = document.querySelector('canvas').getContext('2d');
    let min = {x: -1, y: -1};
    let max = {x: -1, y: -1};
    let rows = text.trim().split('\r\n').map(row => {
      let nums = row.replace(' -> ', ',').split(',')
      let p = {
        x1: parseInt(nums[0]),
        y1: parseInt(nums[1]),
        x2: parseInt(nums[2]),
        y2: parseInt(nums[3]),
      };
      min.x = min.x == -1 ? Math.min(p.x1, p.x2) : Math.min(min.x, p.x1, p.x2);
      min.y = min.y == -1 ? Math.min(p.y1, p.y2) : Math.min(min.y, p.y1, p.y2);
      max.x = max.x == -1 ? Math.max(p.x1, p.x2) : Math.max(max.x, p.x1, p.x2);
      max.y = max.y == -1 ? Math.max(p.y1, p.y2) : Math.max(max.y, p.y1, p.y2);
      return p;
    });

    console.log(min, max)
    ctx.canvas.width = max.x + min.x;
    ctx.canvas.height = max.y + min.y;
    ctx.lineWidth = 2;
    ctx.fillStyle = '#000';
    ctx.fillRect(0, 0, ctx.canvas.width, ctx.canvas.height);
    ctx.strokeStyle = '#fff';
    ctx.globalAlpha = 0.5;
    rows.forEach(line => {
      ctx.beginPath();
      ctx.moveTo(line.x1, line.y1);
      ctx.lineTo(line.x2, line.y2);
      ctx.stroke();
    });
  });
